pkg/telemetry/housekeeper.go

```diff
@@ -0,0 +1,148 @@
+// This file and its contents are licensed under the Apache License 2.0.
+// Please see the included NOTICE for copyright information and
+// LICENSE for a copy of the license.
+
+package telemetry
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/timescale/promscale/pkg/log"
+	"github.com/timescale/promscale/pkg/pgxconn"
+	"github.com/timescale/promscale/pkg/util"
+)
+
+const housekeeperLockId = 0x2D829A932AAFCEDE // Random.
+
+type housekeeper struct {
+	engineCopy    *engineImpl
+	lock          util.AdvisoryLock
+	isHousekeeper bool
+	stop          chan struct{}
+}
+
+// Try tries to become a housekeeper of the current Promscale session.
+func (t *housekeeper) Try() (success bool, err error) {
+	if t.isHousekeeper {
+		// If already a housekeeper, we need to explicitly say that Try was false,
+		// since pg_try_advisory_lock() returns true for the same lockID for that
+		// Promscale session.
+		return false, nil
+	}
+	acquired, err := t.lock.GetAdvisoryLock()
+	if err != nil {
+		return false, fmt.Errorf("attemping telemetry pg-advisory-lock: %w", err)
+	}
+	if acquired {
+		t.isHousekeeper = true
+	}
+	return acquired, nil
+}
+
+func (t *housekeeper) Start() error {
+	if !t.isHousekeeper {
+		return fmt.Errorf("not a housekeeper: cannot do housekeeping")
+	}
+
+	t.stop = make(chan struct{})
+	go t.housekeeping()
+
+	return nil
+}
+
+func (t *housekeeper) Stop() error {
+	t.stop <- struct{}{}
+	close(t.stop)
+	_, err := t.lock.Unlock()
+	if err != nil {
+		return fmt.Errorf("unlock: %w", err)
+	}
+	t.lock.Close()
+	return nil
+}
+
+func (t *housekeeper) Sync() error {
+	informationTableStats, err := getInstanceInformationStats(t.engineCopy.conn)
+	if err != nil {
+		return fmt.Errorf("get instance stats: %w", err)
+	}
+	t.engineCopy.writeToTimescaleMetadataTable(convertStatsToMetadata(informationTableStats))
+	return nil
+}
+
+const housekeepingDuration = time.Hour
+
+func (t *housekeeper) housekeeping() {
```

Comment:
Might be better to call this `run` instead of `housekeeping`. Seems more idiomatic.

Reply:
Since this function will now be on `engineImpl` struct, I think `run` there will be confusing.